New Styling & CSS guide (#204)
withastro/docs: src/pages/en/guides/styling.md

layout: ~/layouts/MainLayout.astro
title: Styling & CSS
description: Learn how to style components with Astro.
setup: |
  import Since from '../../../components/Since.astro';
---
Astro includes special handling to make writing CSS as easy as possible. You can use your favorite libraries, write your own css files, or take advantage of Astro's own scoped styling.

> ## Astro scoped `<style>`

Reviewer: This feels a bit like getting pushed into the deep end. Maybe a section before this that just re-introduces the concept and syntax of styling in Astro components?

- Styling Your Astro Components
- Scoped Styles
  - ...
- Global Styles
  - ...

layout: ~/layouts/MainLayout.astro
title: Styling & CSS
description: Learn how to style components with Astro.
setup: |
  import Since from '../../../components/Since.astro';
---
Astro includes special handling to make writing CSS as easy as possible. You can use your favorite libraries, write your own css files, or take advantage of Astro's own scoped styling.

## Styling an Astro Component

Astro allows you to include both global and scoped styles right inside an Astro component. You can write plain CSS (or use a CSS language extension) directly in Astro's `<style>` tag to style your component.

> By adding an [integration](/en/integrations/integration-guide) like Tailwind, you can also take advantage of writing utility classes inline in your Astro component template.

Reviewer: This is throwing a lot of concepts at the reader for an introduction to styling. Lets see if we can lighten it up.

What about:

```diff
---

Astro was designed for familiar, powerful styling. Write your own CSS directly inside of an Astro component or import your favorite CSS library like [Tailwind](#TODO). Advanced styling languages like [Sass](#TODO) and [Less](#TODO) are also supported.

## Styling in Astro

The easiest way to style your Astro component is with a basic `<style>` tag. When you place a `<style>` tag in your component template, Astro knows to build, optimize and scope your styles for you, automatically.

'''
TODO: SUPER SIMPLE EXAMPLE
Like, just a style tag in a basic component. 
No need to make the distinction between scoping and global just yet.
'''

### Scoped Styles

- Styling inside of Astro components is done by adding a `<style>` tag in the component template.

By default, all Astro component styles are **scoped** // ...
```

description: Learn how to style components with Astro.
---

> INTRO BLAH BLAH: Astro includes special handling to make writing CSS as easy as possible. You can use your favorite libraries, write your own css files, or take advantage of Astro's own scoped styling to target UI frameworks at the component level.

Reviewer 2: Would be cool if I’m wrong, but I’m pretty sure style scoping *only* works for Astro components. There may be ways to get something similar for other frameworks, but I usually think of Astro’s scoping feature as scoping at the `.astro` file level.

Astro automatically minifies and bundles...

## Astro scoped `<style>`

Styling inside of Astro components is done by adding a `<style>` tag in the component template.

> By default, all Astro component styles are **scoped**, meaning they only apply to the HTML elements written in the current component. Any HTML rendered via a child component import is **not** affected by the `<style>` tag.

Reviewer 2: It’s probably OK to leave this like this, but this is not 100% accurate because of how CSS cascades/inherits. Child components won’t be targeted directly by the selectors in `<style>`, but something like the following example would still see the child component inherit a smaller font size.

```astro
<div class="small">
  Small text
  <Component />
</div>

<style>
.small { font-size: 0.5em; }
</style>
```

That said, I can also see the value of the current version for someone who is like “Hey, I _said_ H1s should be red, dammit, why is my child component not red?!”

Author: I *do* want to be accurate! At the same time, we get a lot of "Why isn't my h1 red?!?!"  So, let's think of wording/an example that satisfies both!

</article>
```

This is a great way to style things like blog posts, or documents with CMS-powered content where the contents live outside of Astro. But be careful: components that appear different based on whether or not they have a certain parent component can become difficult to troubleshoot.

## Astro `<style global>`

> To apply styles globally, without the need for a .css file or stylesheet link, Astro allows you to use `<style global>` within any Astro component, page or layout to wrap style definitions. 

Reviewer 2: What do you think of this? Reworked slightly as I found the idea of “wrapping style definitions” slightly confusing.

```suggestion
To apply styles globally, without the need for a separate `.css` file or external stylesheet link, Astro allows you to use the `global` attribute on style declarations within any Astro component, page or layout. 
```

</style>
```

> ⚠️ These global styles apply throughout your entire project! 

It may be easy to lose track of which Astro component is defining styles globally, and harder to troubleshoot errant global styles when they’re scattered around and not in a central CSS file. So, this method is recommended only when you cannot otherwise apply global styling via an import or `<link>`.

## External Styles

There are two ways to resolve global stylesheets: an ESM import for files located within your project source, and an absolute URL link for files in your /public directory, or hosted outside of your project.

### Import a Global Stylesheet

> Astro detects these CSS imports and then builds, optimizes, and adds the CSS to the page automatically. 

Reviewer 2: I can see why you might lead with this line as it advertises some positive things, but logically I think it might make more sense just after the example. Otherwise maybe needs a bit of work as 

```
Import a Global Stylesheet

...these imports
```

isn’t clear about what “these” imports are.

Import a global stylesheet **at the top** of an Astro component script, using its relative file path, along with any other imports:

```astro
---
// Astro will include and optimize this CSS for you automatically
// This also works for preprocessor files like .scss, .styl, etc.
import '../styles/utils.css';

const title="My Astro Page"
---
<html><!-- Your page here --></html>
```

Importing CSS files should work anywhere that ESM imports are supported, including:
- JavaScript files
- TypeScript files
- Astro component front matter
- non-Astro components like React, Svelte, and others

When a CSS file is imported using this method, any `@import` statements are also resolved and inlined into the imported CSS file.

### Stylesheet Link

When your stylesheet is located in your `/public` directory, or when using a public stylesheet hosted offsite (e.g. a Prism theme), use a `<link>` tag with an absolute URL reference.

```html
<head>

>   <link rel="stylesheet" href="/styles/global.css" />

Reviewer 2: Would it be worth splitting out two subsections, one for external URLs, one for files in `/public`? That might seem excessive, but a brief note on how `/public/style.css` is served at `/style.css` might be helpful.

Author: In a new PR, I've now added a section in "Static Assets" that describes using files that are in `public/`, So I'll look over this again, with that in mind, and see whether a note/link to that page might do the trick.

  <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/prismjs@1.24.1/themes/prism-tomorrow.css">
</head>
```

This approach skips the CSS processing, bundling and optimizations that are provided by Astro when you use the `import` method described above. These files will not be transformed.

## Variables in Styles

> In Astro v0.21+, _serializable_ server-side variables can be passed into client-side `<style>`.

Reviewer 2: Is this an opportunity to use `<Since>` 👀 

```suggestion
<Since v="0.21.0" />

_Serializable_ server-side variables can be passed into client-side `<style>`.
```

<h-tick>✓</h-tick>
```

## PostCSS

You can use Tailwind or any [PostCSS plugin](https://www.postcss.parts/) by adding a `postcss.config.cjs` file to the root of your project. Follow the documentation for the plugin you’re trying to install for configuration and setup.

Be aware that this plugin will run on all CSS in your project, including any files that compiled to CSS (like `.scss` Sass files, for example).

_Note: CSS in `public/` **will not be transformed!** Instead, place it within `src/` if you’d like PostCSS to run over your styles._

### Autoprefixer

> [Autoprefixer][autoprefixer] is a PostCSS plugin that takes care of cross-browser CSS compatibility for you. Use it in astro by installing it (`npm install --save-dev autoprefixer`) and adding a `postcss.config.cjs` file to the root of your project:

Reviewer 2: Nit: capitalise Astro

```suggestion
[Autoprefixer][autoprefixer] is a PostCSS plugin that takes care of cross-browser CSS compatibility for you. Use it in Astro by installing it (`npm install --save-dev autoprefixer`) and adding a `postcss.config.cjs` file to the root of your project:
```

<h-tick>✓</h-tick>
```

## PostCSS

You can use Tailwind or any [PostCSS plugin](https://www.postcss.parts/) by adding a `postcss.config.cjs` file to the root of your project. Follow the documentation for the plugin you’re trying to install for configuration and setup.

Be aware that this plugin will run on all CSS in your project, including any files that compiled to CSS (like `.scss` Sass files, for example).

>💡 _CSS in `public/` **will not be transformed!** Instead, place it within `src/` if you’d like PostCSS to run over your styles._

> ### Autoprefixer

Reviewer: similar to Tailwind, I think this section can be removed as "too opinionated". If someone wants to add autoprefixer, then the docs above is enough to follow to add it.

If you want to keep a reference, then the PostCSS section could be rewritten to:

```
Astro supports [PostCSS](https://www.postcss.parts/) out of the box. To enable, add a `postcss.config.cjs` file to the root of your project and add your (like autoprefixer)  to the file. Astro will automatically run your styles through PostCSS.

};
```

> ### Tailwind

Reviewer: this section can be removed now, and replaced with a reference/link to the new integration README / guides.

</head>
```

#### Migrating to v0.24

> As of [version 0.24.0-next](https://github.com/withastro/astro/releases/tag/astro%400.24.0-next.0) Astro will no longer support [Astro.resolve()][astro-resolve]. See the [Migration Guide](/en/migrate) for more details.

Reviewer 2: I guess this should be updated:

```suggestion
As of [version 0.24.0](https://github.com/withastro/astro/releases/tag/astro%400.24.0), Astro no longer supports [Astro.resolve()][astro-resolve]. See the [Migration Guide](/en/migrate) for more details.
```

## CSS Preprocessors

Astro supports CSS preprocessors such as [Sass][sass], [Stylus][stylus], and [Less][less] through [Vite][vite-preprocessors].

### Sass

 ```
 npm install -D sass
 ````

>  Use  `<style lang="scss">` or `<style lang="sass">` (indented) in `.astro` files

Reviewer 2: What does the “indented” mean here? Should I indent `<style lang="sass">`?

Author: I don't know! I'm not sure where I pulled that from, but I'll check on it....

### Stylus

```
npm install -D stylus
``` 
Use `<style lang="styl">` or `<style lang="stylus">` in `.astro` files

### Less
```
npm install -D less
``` 
Use `<style lang="less">` in `.astro` files.

> You can also use all of the above CSS preprocessors within JS frameworks as well! Be sure to follow the patterns each framework recommends:

- **React** / **Preact**: `import Styles from './styles.module.scss'`;
- **Vue**: `<style lang="scss">`
- **Svelte**: `<style lang="scss">`

> Additionally, [PostCSS](#-postcss) is supported, but the setup is [slightly different](#-postcss).

Reviewer 2: Broken link?

Author: Ah, I reorganized this page, so yes, an internal link based on headers would have changed.

@tailwind utilities;
```

4. Import your stylesheet to your Astro page (or layout template):

```astro
---
import '../styles/global.css';
---
```

> #### Migrating to v0.24

Reviewer: no longer relevant!
